runtime/src/main/java/com/dylibso/chicory/runtime/TypeValidator.java

```diff
@@ -155,140 +175,102 @@ private static ValueType getLocalType(List<ValueType> localTypes, int idx) {
     public void validate(FunctionBody body, FunctionType functionType, Instance instance) {
         var localTypes = body.localTypes();
         var inputLen = functionType.params().size();
-        push(prevStack, clone(valueTypeStack));
-        push(returns, functionType.returns());
+        pushCtrl(null, new ArrayList<>(), functionType.returns());
 
         for (var i = 0; i < body.instructions().size(); i++) {
             var op = body.instructions().get(i);
 
             // control flow instructions handling
             switch (op.opcode()) {
-                case LOOP:
-                case BLOCK:
+                case UNREACHABLE:
                     {
-                        push(prevStack, clone(valueTypeStack));
-                        push(returns, getReturns(op, instance));
+                        unreachable();
                         break;
                     }
                 case IF:
                     {
-                        popAndVerifyType(ValueType.I32);
-                        push(prevStack, clone(valueTypeStack));
+                        popVal(ValueType.I32);
+                        // fallthrough
+                    }
+                case LOOP: // t1* -> t2*
```

Comment:
```suggestion
                case LOOP: // t1* -> t2*
                     // fallthrough
```

Reply:
In this case the comment is a placeholder instead of `break`, any reason to push it down?

Reply:
Oh, I am not suggesting to push it down, I am suggesting to add it on LOOP too, since its body is empty